File: reference/typescript/src/runtime/api-server.js
import http from "node:http";

export function startApiServer(service, options) {
  const base = options.path ?? "/aep/api";
  const server = http.createServer((req, res) => {
    handle(service, base, req, res).catch((err) => {
      sendJson(res, 500, { error: err.message });
    });
  });
  return new Promise((resolve) => {
    server.listen(options.port ?? 0, options.host ?? "127.0.0.1", () => {
      const addr = server.address();
      resolve({
        port: addr.port,
        stop: () => new Promise((done) => server.close(done))
      });
    });
  });
}

async function handle(service, base, req, res) {
  const url = new URL(req.url, `http://${req.headers.host}`);
  const route = url.pathname.startsWith(base) ? url.pathname.slice(base.length) : null;

  if (route === "/healthz" && req.method === "GET") {
    return sendJson(res, 200, {
      status: "ok",
      runtime: service.config.runtime,
      delivery: await service.getStats()
    });
  }

  if (route === "/events" && req.method === "POST") {
    return handleIngest(service, req, res);
  }

  if (route === "/dlq" && req.method === "GET") {
    const records = await service.getDeadLettered();
    return sendJson(res, 200, { deadLettered: records.length, records });
  }

  if (route === "/pending" && req.method === "GET") {
    const records = await service.getPending();
    return sendJson(res, 200, { pending: records.length, records });
  }

  if (route === "/stats" && req.method === "GET") {
    return sendJson(res, 200, await service.getStats());
  }

  if (route === "/subscriptions" && req.method === "POST") {
    return handleCreateSubscription(service, req, res);
  }

  if (route === "/subscriptions" && req.method === "GET") {
    return sendJson(res, 200, { subscriptions: service.listSubscriptions() });
  }

  const subMatch = route && route.match(/^\/subscriptions\/([^/]+)(\/events|\/stream)?$/);
  if (subMatch) {
    const id = decodeURIComponent(subMatch[1]);
    const suffix = subMatch[2];
    if (!suffix && req.method === "GET") {
      const record = service.getSubscription(id);
      return record ? sendJson(res, 200, record) : sendJson(res, 404, { error: "not found" });
    }
    if (!suffix && req.method === "DELETE") {
      const deleted = await service.deleteSubscription(id);
      return deleted ? sendJson(res, 200, { deleted: true }) : sendJson(res, 404, { error: "not found" });
    }
    if (suffix === "/events" && req.method === "GET") {
      return handleLongPoll(service, id, res);
    }
    if (suffix === "/stream" && req.method === "GET") {
      return handleStream(service, id, req, res);
    }
  }

  return sendJson(res, 404, { error: "not found" });
}

async function handleIngest(service, req, res) {
  const chunks = [];
  for await (const chunk of req) chunks.push(chunk);
  const raw = Buffer.concat(chunks).toString("utf8");
  let event;
  try {
    event = JSON.parse(raw);
  } catch {
    return sendJson(res, 400, { accepted: false, errors: ["invalid JSON body"] });
  }
  try {
    service.publish(event);
  } catch (err) {
    const errors = err.message.startsWith("invalid AEP event: ")
      ? err.message.slice("invalid AEP event: ".length).split("; ")
      : [err.message];
    return sendJson(res, 400, { accepted: false, errors });
  }
  return sendJson(res, 202, { accepted: true, id: event.id });
}

function sendJson(res, status, body) {
  res.writeHead(status, { "Content-Type": "application/json" });
  res.end(JSON.stringify(body));
}

async function handleCreateSubscription(service, req, res) {
  const chunks = [];
  for await (const chunk of req) chunks.push(chunk);
  const raw = Buffer.concat(chunks).toString("utf8");
  let body;
  try {
    body = raw.length > 0 ? JSON.parse(raw) : {};
  } catch {
    return sendJson(res, 400, { error: "invalid JSON body" });
  }
  const filter = body.filter ?? body;
  const record = await service.createSubscription(filter);
  return sendJson(res, 201, record);
}

function handleLongPoll(service, id, res) {
  if (!service.getSubscription(id)) return sendJson(res, 404, { error: "not found" });
  const events = service.takeEvents(id, 100);
  return sendJson(res, 200, { events });
}

function handleStream(service, id, req, res) {
  if (!service.getSubscription(id)) return sendJson(res, 404, { error: "not found" });
  res.writeHead(200, {
    "Content-Type": "text/event-stream",
    "Cache-Control": "no-cache",
    Connection: "keep-alive"
  });
  const buffered = service.takeEvents(id, 1000);
  for (const evt of buffered) res.write(`data: ${JSON.stringify(evt)}\n\n`);
  const detach = service.attachStream(id, (evt) => {
    res.write(`data: ${JSON.stringify(evt)}\n\n`);
  });
  req.on("close", () => { if (detach) detach(); });
}

File: implementations/java/src/main/java/com/axisrobo/aep/runtime/ApiServer.java
package com.axisrobo.aep.runtime;

import com.axisrobo.aep.Envelope;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Executors;

public class ApiServer {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final HarmovelaRuntimeService service;
    private final String base;
    private HttpServer server;
    private int port;

    public ApiServer(HarmovelaRuntimeService service, Config.Transport t) {
        this.service = service;
        this.base = t.path() == null || t.path().isEmpty() ? "/harmovela/api" : t.path();
    }

    public int port() { return port; }

    public void start(String host, int requestedPort) throws IOException {
        server = HttpServer.create(new InetSocketAddress(host, requestedPort), 0);
        server.createContext(base, this::handle);
        server.setExecutor(Executors.newCachedThreadPool());
        server.start();
        port = server.getAddress().getPort();
    }

    public void stop() {
        if (server != null) {
            server.stop(0);
            server = null;
        }
    }

    private void handle(HttpExchange exchange) throws IOException {
        var path = exchange.getRequestURI().getPath();
        var route = path.startsWith(base) ? path.substring(base.length()) : path;
        var method = exchange.getRequestMethod();
        try {
            if (route.equals("/healthz") && method.equals("GET")) {
                send(exchange, 200, Map.of(
                    "status", "ok",
                    "runtime", Map.of("id", service.config().runtimeId(), "source", service.config().runtimeSource()),
                    "delivery", service.getStats()));
            } else if (route.equals("/events") && method.equals("POST")) {
                handleIngest(exchange);
            } else if (route.equals("/dlq") && method.equals("GET")) {
                List<Map<String, Object>> records = service.getDeadLettered();
                send(exchange, 200, Map.of("deadLettered", records.size(), "records", records));
            } else if (route.equals("/pending") && method.equals("GET")) {
                List<Map<String, Object>> records = service.getPending();
                send(exchange, 200, Map.of("pending", records.size(), "records", records));
            } else if (route.equals("/stats") && method.equals("GET")) {
                send(exchange, 200, service.getStats());
            } else if (route.equals("/subscriptions") && method.equals("POST")) {
                handleCreateSubscription(exchange);
            } else if (route.equals("/subscriptions") && method.equals("GET")) {
                send(exchange, 200, Map.of("subscriptions", service.listSubscriptions()));
            } else if (route.startsWith("/subscriptions/")) {
                handleSubscriptionItem(route, method, exchange);
            } else {
                send(exchange, 404, Map.of("error", "not found"));
            }
        } catch (Exception e) {
            send(exchange, 500, Map.of("error", String.valueOf(e.getMessage())));
        }
    }

    @SuppressWarnings("unchecked")
    private void handleIngest(HttpExchange exchange) throws IOException {
        var raw = new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8);
        Map<String, Object> event;
        try {
            event = (Map<String, Object>) MAPPER.readValue(raw, Map.class);
        } catch (Exception e) {
            send(exchange, 400, Map.of("accepted", false, "errors", List.of("invalid JSON body")));
            return;
        }
        var errors = Envelope.validate(event);
        if (!errors.isEmpty()) {
            send(exchange, 400, Map.of("accepted", false, "errors", errors));
            return;
        }
        service.publish(event);
        send(exchange, 202, Map.of("accepted", true, "id", event.get("id")));
    }

    private void send(HttpExchange exchange, int status, Map<String, Object> body) throws IOException {
        var payload = MAPPER.writeValueAsBytes(body);
        exchange.getResponseHeaders().set("Content-Type", "application/json");
        exchange.sendResponseHeaders(status, payload.length);
        try (var os = exchange.getResponseBody()) {
            os.write(payload);
        }
    }

    @SuppressWarnings("unchecked")
    private void handleCreateSubscription(HttpExchange exchange) throws IOException {
        var raw = new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8);
        Map<String, Object> body;
        try {
            body = raw.isEmpty() ? Map.of() : (Map<String, Object>) MAPPER.readValue(raw, Map.class);
        } catch (Exception e) {
            send(exchange, 400, Map.of("error", "invalid JSON body"));
            return;
        }
        Map<String, Object> filter = body.get("filter") instanceof Map<?, ?> f
            ? (Map<String, Object>) f : body;
        var record = service.createSubscription(filter);
        send(exchange, 201, record);
    }

    private void handleSubscriptionItem(String route, String method, HttpExchange exchange) throws IOException {
        var rest = route.substring("/subscriptions/".length());
        if (rest.endsWith("/events") && method.equals("GET")) {
            var id = rest.substring(0, rest.length() - "/events".length());
            if (service.getSubscription(id) == null) { send(exchange, 404, Map.of("error", "not found")); return; }
            send(exchange, 200, Map.of("events", service.takeEvents(id, 100)));
            return;
        }
        if (rest.endsWith("/stream") && method.equals("GET")) {
            handleStream(rest.substring(0, rest.length() - "/stream".length()), exchange);
            return;
        }
        if (rest.contains("/")) { send(exchange, 404, Map.of("error", "not found")); return; }
        if (method.equals("GET")) {
            var record = service.getSubscription(rest);
            if (record == null) send(exchange, 404, Map.of("error", "not found"));
            else send(exchange, 200, record);
        } else if (method.equals("DELETE")) {
            if (service.deleteSubscription(rest)) send(exchange, 200, Map.of("deleted", true));
            else send(exchange, 404, Map.of("error", "not found"));
        } else {
            send(exchange, 404, Map.of("error", "not found"));
        }
    }

    private void handleStream(String id, HttpExchange exchange) throws IOException {
        if (service.getSubscription(id) == null) { send(exchange, 404, Map.of("error", "not found")); return; }
        exchange.getResponseHeaders().set("Content-Type", "text/event-stream");
        exchange.getResponseHeaders().set("Cache-Control", "no-cache");
        exchange.sendResponseHeaders(200, 0);
        var os = exchange.getResponseBody();
        var lock = new Object();
        try {
            writeRaw(os, lock, ": ok\n\n");
            for (var evt : service.takeEvents(id, 1000)) writeSse(os, lock, evt);
            var queue = new java.util.concurrent.LinkedBlockingQueue<Map<String, Object>>();
            var detach = service.attachStream(id, queue::offer);
            try {
                while (true) {
                    var evt = queue.take();
                    writeSse(os, lock, evt);
                }
            } finally {
                if (detach != null) detach.run();
            }
        } catch (Exception ignored) {
        } finally {
            try { os.close(); } catch (IOException ignored) {}
        }
    }

    private void writeSse(OutputStream os, Object lock, Map<String, Object> evt) throws IOException {
        writeRaw(os, lock, "data: " + MAPPER.writeValueAsString(evt) + "\n\n");
    }

    private void writeRaw(OutputStream os, Object lock, String text) throws IOException {
        synchronized (lock) {
            os.write(text.getBytes(StandardCharsets.UTF_8));
            os.flush();
        }
    }
}
